src/Illuminate/Foundation/Testing/Concerns/InteractsWithGate.php

```diff
@@ -0,0 +1,142 @@
+<?php
+
+namespace Illuminate\Foundation\Testing\Concerns;
+
+use Illuminate\Foundation\Testing\ExpectedPermission;
+use Illuminate\Support\Facades\Gate;
+
+trait InteractsWithGate
+{
+    /**
+     * Indicates if the hooks have been registered.
+     *
+     * @var bool $hasRegisteredGateHooks
+     */
+    private $hasRegisteredGateHooks = false;
+
+    /**
+     * Collection of expected permissions.
+     *
+     * @var \Illuminate\Support\Collection
+     */
+    private $expectedPermissions;
+
+    /**
+     * Register the gate hooks and expectations. Also initializes the expected permissions collection.
+     *
+     * @return void
+     */
+    private function registerGateHooks()
+    {
+        if ($this->hasRegisteredGateHooks) {
+            return;
+        }
+
+        $this->expectedPermissions = collect();
+
+        Gate::before(function ($user, $ability, $arguments) {
+            if ($permission = $this->expectedPermission($user, $ability, $arguments)) {
+                return $permission->expectedResult();
+            }
+
+            return null;
+        });
+
+        $this->beforeApplicationDestroyed(function () {
+            $unmatched = $this->expectedPermissions->reject(function ($permision) {
+                return $permision->wasMatched();
+            })->map(function ($permission) {
+                return $permission->ability();
+            });
+
+            if ($unmatched->isNotEmpty()) {
+                $this->fail(sprintf('The expected ability or policiy checks were not called: %s', $unmatched->join(', ')));
+            }
+        });
+
+        $this->hasRegisteredGateHooks = true;
+    }
+
+    /**
+     * Gets the expected permission from the list of expected permissions.
+     *
+     * @param \Illuminate\Contracts\Auth\Authenticatable $user
+     * @param string $ability
+     * @param array $arguments
```

Comment:
Please use the same conventions as the rest of the framework. Please also adjust other docblocks in this PR.

```suggestion
     * @param  \Illuminate\Contracts\Auth\Authenticatable  $user
     * @param  string  $ability
     * @param  array  $arguments
```